README.rst
==========================
repoze.who.plugins.vepauth
==========================

This is a repoze.who plugin for automated authentication via BrowserID:

    https://browserid.org/
    https://wiki.mozilla.org/Identity/BrowserIDSync

The plugin implements an experimental protocol for authenticating to ReSTful
web services with the Verified Email Protocol, a.k.a Mozilla's BrowserID
project.  It is designed for use in automated tools like the Firefox Sync
Client.  If you're looking for something to use for human visitors on your
site, please try:

    http://github.com/mozilla-services/repoze.who.plugins.browserid

When accessing a protected resource, the server will generate a 401 challenge
response with the scheme "OAuth+VEP" as follows::

    > GET /protected_resource HTTP/1.1
    > Host: example.com

    < HTTP/1.1 401 Unauthorized
    < WWW-Authenticate: OAuth+VEP url="/request_token"

The client should extract the url from this challenge and POST a VEP assertion
to that location.  This will create a new authentication session and return a
set of OAuth client credentials::

    > POST /request_token HTTP/1.1
    > Host: example.com
    > Content-Type: application/x-www-form-urlencoded
    >
    > assertion=VEP_ASSERTION_DATA

    < HTTP/1.1 200 OK
    < Content-Type: application/json
    <
    < {
    <   "oauth_consumer_key": SESSION_TOKEN,
    <   "oauth_consumer_secret": SESSION_SECRET
    < }

Subsequent requests should be signed using these credentials in Two-Legged
OAuth mode::

    > GET /protected_resource HTTP/1.1
    > Host: example.com
    > Authorization: OAuth oauth_consumer_key=SESSION_TOKEN,
    >                      oauth_signature_method="HMAC-SHA1",
    >                      oauth_version="1.0",
    >                      oauth_timestamp=TIMESTAMP,
    >                      oauth_nonce=NONCE
    >                      oauth_signature=SIGNATURE

    < HTTP/1.1 200 OK
    < Content-Type: text/plain
    <
    < For your eyes only:  secret data!

Session tokens are timestamped and will eventually expire.  If this happens
you will receive a 401 response as before, and should POST a new assertion
to obtain fresh credentials.

Extending the token management
------------------------------

`repoze.who.plugins.vepauth` is extensible. If you want to provide a different
mechanism to manage the tokens, you can do so by providing a different token
manager instance to the plugin with the `token_manager` option.

The `TokenManager` class have two methods than need to be implemented (it's an
abstract class): `make_token` and `parse_token`. The implementation details are
left to the childs classes.

`repose.who.plugins.vepauth` comes with one `SignedTokenManager` which
implement a simple token management class in pure python. It has a number of
methods that can be overridden to customize its behavior.


CHANGES.txt
0.3.0 - 2012-02-14
==================

 * Replaced Two-Legged OAuth with MAC Access Auth, implemented according
   to the latest draft standard:

      https://tools.ietf.org/html/draft-ietf-oauth-v2-http-mac-01


0.2.0 - 2012-02-14
==================

 * Added basic pattern-matching for the token_url, e.g. "/{app}/token".
 * Made TokenManager.make_token and TokenManager.parse_token accept the
   incoming request as first argument, mostly so they can get at the
   results of pattern-matching.
 * Let TokenManager.make_token return a dict of extra info to be included
   in the response.


0.1.0 - 2012-02-07
==================

 * Initial release.


repoze/who/plugins/vepauth/tests/__init__.py
# ***** BEGIN LICENSE BLOCK *****
# Version: MPL 1.1/GPL 2.0/LGPL 2.1
#
# The contents of this file are subject to the Mozilla Public License Version
# 1.1 (the "License"); you may not use this file except in compliance with
# the License. You may obtain a copy of the License at
# http://www.mozilla.org/MPL/
#
# Software distributed under the License is distributed on an "AS IS" basis,
# WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
# for the specific language governing rights and limitations under the
# License.
#
# The Original Code is repoze.who.plugins.vepauth
#
# The Initial Developer of the Original Code is the Mozilla Foundation.
# Portions created by the Initial Developer are Copyright (C) 2011
# the Initial Developer. All Rights Reserved.
#
# Contributor(s):
#   Ryan Kelly (rkelly@mozilla.com)
#
# Alternatively, the contents of this file may be used under the terms of
# either the GNU General Public License Version 2 or later (the "GPL"), or
# the GNU Lesser General Public License Version 2.1 or later (the "LGPL"),
# in which case the provisions of the GPL or the LGPL are applicable instead
# of those above. If you wish to allow use of your version of this file only
# under the terms of either the GPL or the LGPL, and not to allow others to
# use your version of this file under the terms of the MPL, indicate your
# decision by deleting the provisions above and replace them with the notice
# and other provisions required by the GPL or the LGPL. If you do not delete
# the provisions above, a recipient may use your version of this file under
# the terms of any one of the MPL, the GPL or the LGPL.
#
# ***** END LICENSE BLOCK *****


repoze/who/plugins/vepauth/tests/test_utils.py
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this file,
# You can obtain one at http://mozilla.org/MPL/2.0/.

import unittest2

from webob import Request

from repoze.who.plugins.vepauth.utils import (strings_differ,
                                              parse_authz_header,
                                              sign_request,
                                              get_mac_signature,
                                              check_mac_signature,
                                              get_normalized_request_string)


class TestUtils(unittest2.TestCase):

    def test_strings_differ(self):
        # We can't really test the timing-invariance, but
        # we can test that we actually compute equality!
        self.assertTrue(strings_differ("", "a"))
        self.assertTrue(strings_differ("b", "a"))
        self.assertTrue(strings_differ("cc", "a"))
        self.assertTrue(strings_differ("cc", "aa"))
        self.assertFalse(strings_differ("", ""))
        self.assertFalse(strings_differ("D", "D"))
        self.assertFalse(strings_differ("EEE", "EEE"))

    def test_parse_authz_header(self):
        def req(authz):
            """Make a fake request with the given authz header."""
            class request:
                environ = {"HTTP_AUTHORIZATION": authz}
            return request

        # Test parsing of a single unquoted parameter.
        params = parse_authz_header(req('Digest realm=hello'))
        self.assertEquals(params['scheme'], 'Digest')
        self.assertEquals(params['realm'], 'hello')

        # Test parsing of multiple parameters with mixed quotes.
        params = parse_authz_header(req('Digest test=one, again="two"'))
        self.assertEquals(params['scheme'], 'Digest')
        self.assertEquals(params['test'], 'one')
        self.assertEquals(params['again'], 'two')

        # Test parsing of an escaped quote and empty string.
        params = parse_authz_header(req('Digest test="\\"",again=""'))
        self.assertEquals(params['scheme'], 'Digest')
        self.assertEquals(params['test'], '"')
        self.assertEquals(params['again'], '')

        # Test parsing of embedded commas, escaped and non-escaped.
        params = parse_authz_header(req('Digest one="1\\,2", two="3,4"'))
        self.assertEquals(params['scheme'], 'Digest')
        self.assertEquals(params['one'], '1,2')
        self.assertEquals(params['two'], '3,4')

        # Test parsing on various malformed inputs
        self.assertRaises(ValueError, parse_authz_header, req(None))
        self.assertRaises(ValueError, parse_authz_header, req(""))
        self.assertRaises(ValueError, parse_authz_header, req(" "))
        self.assertRaises(ValueError, parse_authz_header,
                          req('Broken raw-token'))
        self.assertRaises(ValueError, parse_authz_header,
                          req('Broken realm="unclosed-quote'))
        self.assertRaises(ValueError, parse_authz_header,
                          req('Broken realm=unopened-quote"'))
        self.assertRaises(ValueError, parse_authz_header,
                          req('Broken realm="unescaped"quote"'))
        self.assertRaises(ValueError, parse_authz_header,
                          req('Broken realm="escaped-end-quote\\"'))
        self.assertRaises(ValueError, parse_authz_header,
                          req('Broken realm="duplicated",,what=comma'))

        # Test all those again, but returning a default value
        self.assertEquals(None, parse_authz_header(req(None), None))
        self.assertEquals(None, parse_authz_header(req(""), None))
        self.assertEquals(None, parse_authz_header(req(" "), None))
        self.assertEquals(None,
                          parse_authz_header(req('Broken raw-token'), None))
        self.assertEquals(None, parse_authz_header(
                          req('Broken realm="unclosed-quote'), None))
        self.assertEquals(None, parse_authz_header(
                          req('Broken realm=unopened-quote"'), None))
        self.assertEquals(None, parse_authz_header(
                          req('Broken realm="unescaped"quote"'), None))
        self.assertEquals(None, parse_authz_header(
                          req('Broken realm="escaped-end-quote\\"'), None))
        self.assertEquals(None, parse_authz_header(
                          req('Broken realm="duplicated",,what=comma'), None))

    def test_normalized_request_string_against_example_from_spec(self):
        # This is the example used in Section 3.2.1 of RFC-TODO
        req = "POST /request?b5=%3D%253D&a3=a&c%40=&a2=r%20b&c2&a3=2+q"\
              " HTTP/1.1\r\n"\
              "Host: example.com\r\n"\
              "\r\n"\
              "Hello World!"
        params = {
            "ts": "264095",
            "nonce": "7d8f3e4a",
            "ext": "a,b,c",
        }
        sigstr = "264095\n"\
                 "7d8f3e4a\n"\
                 "POST\n"\
                 "/request?b5=%3D%253D&a3=a&c%40=&a2=r%20b&c2&a3=2+q\n"\
                 "example.com\n"\
                 "80\n"\
                 "a,b,c\n"
        # IanB, *thank you* for Request.from_string!
        req = Request.from_string(req)
        mysigstr = get_normalized_request_string(req, params)
        self.assertEquals(sigstr, mysigstr)

    def test_get_mac_signature_against_example_from_spec(self):
        # This is the example used in Section 1.1 of RFC-TODO
        req = "GET /resource/1?b=1&a=2 HTTP/1.1\r\n"\
              "Host: example.com\r\n"\
              "\r\n"
        params = {
            "id": "h480djs93hd8",
            "ts": "1336363200",
            "nonce": "dj83hs9s"
        }
        secret = "489dks293j39"
        sigstr = "1336363200\n"\
                 "dj83hs9s\n"\
                 "GET\n"\
                 "/resource/1?b=1&a=2\n"\
                 "example.com\n"\
                 "80\n"\
                 "\n"
        sig = "bhCQXTVyfj5cmA9uKkPFx1zeOXM="
        req = Request.from_string(req)
        mysigstr = get_normalized_request_string(req, params)
        self.assertEquals(sigstr, mysigstr)
        mysig = get_mac_signature(req, secret, params)
        # XXX: disagrees with spec, but I'm wondering if spec is broken..?
        # self.assertEquals(sig, mysig)

    def test_sign_request_throws_away_other_auth_params(self):
        req = Request.blank("/")
        req.authorization = ("Digest", {"response": "helloworld"})
        sign_request(req, "token", "secret")
        self.assertEquals(req.authorization[0], "MAC")

    def test_normalized_request_string_with_custom_port(self):
        req = "GET / HTTP/1.1\r\nHost: example.com:88\r\n\r\n"
        req = Request.from_string(req)
        req.authorization = ("MAC", {"ts": "1", "nonce": "2"})
        sigstr = "1\n2\nGET\n/\nexample.com\n88\n\n"
        mysigstr = get_normalized_request_string(req)
        self.assertEquals(sigstr, mysigstr)

    def test_normalized_request_string_with_https_scheme(self):
        req = "GET / HTTP/1.1\r\nHost: example.com\r\n\r\n"
        req = Request.from_string(req)
        req.authorization = ("MAC", {"ts": "1", "nonce": "2"})
        req.scheme = "https"
        sigstr = "1\n2\nGET\n/\nexample.com\n443\n\n"
        mysigstr = get_normalized_request_string(req)
        self.assertEquals(sigstr, mysigstr)

    def test_normalized_request_string_errors_when_no_default_port(self):
        req = "GET / HTTP/1.1\r\nHost: example.com\r\n\r\n"
        req = Request.from_string(req)
        req.authorization = ("MAC", {"ts": "1", "nonce": "2"})
        req.scheme = "httptypo"
        self.assertRaises(ValueError, get_normalized_request_string, req)

    def test_check_mac_signature_errors_when_missing_data(self):
        req = "GET / HTTP/1.1\r\nHost: example.com\r\n\r\n"
        req = Request.from_string(req)
        req.authorization = ("MAC", {"ts": "1", "nonce": "2"})
        self.assertFalse(check_mac_signature(req, "secretkeyohsecretkey"))

    def test_compatability_with_ff_sync_client(self):
        # These are test values used in the FF Sync Client testsuite.
        # Trying to make sure we're compatible.
        token, secret = (
          "vmo1txkttblmn51u2p3zk2xiy16hgvm5ok8qiv1yyi86ffjzy9zj0ez9x6wnvbx7",
          "b8u1cc5iiio5o319og7hh8faf2gi5ym4aq0zwf112cv1287an65fudu5zj7zo7dz",
        )
        req = "GET /alias/ HTTP/1.1\r\nHost: 10.250.2.176\r\n\r\n"
        req = Request.from_string(req)
        req.authorization = ("MAC", {"ts": "1329181221", "nonce": "wGX71"})
        sig = "jzh5chjQc2zFEvLbyHnPdX11Yck="
        mysig = get_mac_signature(req, secret)
        self.assertEquals(sig, mysig)


repoze/who/plugins/vepauth/tests/test_nonces.py
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this file,
# You can obtain one at http://mozilla.org/MPL/2.0/.

import unittest2
import time

from repoze.who.plugins.vepauth.noncemanager import NonceManager, Cache


class TestNonces(unittest2.TestCase):

    def test_default_timeout_is_one_minute(self):
        nm = NonceManager()
        self.assertEquals(nm.nonce_timeout,  60)
        self.assertEquals(nm.token_timeout,  60)

    def test_operation(self):
        timeout = 0.1
        nm = NonceManager(nonce_timeout=timeout, token_timeout=1)
        # Initially nothing is cached, so all nonces as fresh.
        self.assertEquals(nm.nonce_timeout, 0.1)
        self.assertEquals(len(nm), 0)
        self.assertTrue(nm.is_fresh("token", time.time(), "abc"))
        # After adding a nonce, it should contain just that item.
        nm.add_nonce("token", time.time(), "abc")
        self.assertEquals(len(nm), 1)
        self.assertFalse(nm.is_fresh("token", time.time(), "abc"))
        self.assertTrue(nm.is_fresh("token", time.time(), "xyz"))
        # After the timeout passes, the nonce should be expired.
        time.sleep(timeout)
        self.assertTrue(nm.is_fresh("token", time.time(), "abc"))
        # Writing to the cache purges expired nonces but keeps valid ones.
        nm.add_nonce("token", time.time(), "abc")
        time.sleep(timeout / 2)
        nm.add_nonce("token", time.time(), "def")
        self.assertFalse(nm.is_fresh("token", time.time(), "abc"))
        self.assertFalse(nm.is_fresh("token", time.time(), "def"))
        self.assertTrue(nm.is_fresh("token", time.time(), "xyz"))
        time.sleep(timeout / 2)
        nm.add_nonce("token", time.time(), "xyz")
        self.assertTrue(nm.is_fresh("token", time.time(), "abc"))
        self.assertFalse(nm.is_fresh("token", time.time(), "def"))
        self.assertFalse(nm.is_fresh("token", time.time(), "xyz"))
        self.assertEquals(len(nm), 2)

    def test_that_cache_items_are_ungettable_once_expired(self):
        timeout = 0.1
        cache = Cache(timeout)
        cache.set("hello", "world")
        self.assertEquals(cache.get("hello"), "world")
        time.sleep(timeout / 2)
        self.assertEquals(cache.get("hello"), "world")
        time.sleep(timeout / 2)
        self.assertRaises(KeyError, cache.get, "hello")

    def test_that_cache_respects_max_size(self):
        timeout = 0.1
        cache = Cache(timeout, max_size=2)
        cache.set("hello", "world")
        self.assertEquals(len(cache), 1)
        cache.set("how", "are")
        self.assertEquals(len(cache), 2)
        cache.set("you", "today?")
        self.assertEquals(len(cache), 2)
        self.assertEquals(cache.get("you"), "today?")
        self.assertEquals(cache.get("how"), "are")
        self.assertRaises(KeyError, cache.get, "hello")


repoze/who/plugins/vepauth/__init__.py
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this file,
# You can obtain one at http://mozilla.org/MPL/2.0/.
"""

A repoze.who plugin for automated authentication via BrowserID:

    https://browserid.org/
    https://wiki.mozilla.org/Identity/BrowserIDSync


This plugin implements an experimental protocol for authenticating to ReSTful
web services with the Verified Email Protocol, a.k.a Mozilla's BrowserID
project.  It is designed for use in automated tools like the Firefox Sync
Client.  If you're looking for something to use for human visitors on your
site, please try:

    http://github.com/mozilla-services/repoze.who.plugins.browserid

"""

__ver_major__ = 0
__ver_minor__ = 3
__ver_patch__ = 0
__ver_sub__ = ""
__ver_tuple__ = (__ver_major__, __ver_minor__, __ver_patch__, __ver_sub__)
__version__ = "%d.%d.%d%s" % __ver_tuple__


import re
import json
import fnmatch

from zope.interface import implements

from webob import Request as Request_, Response

from repoze.who.interfaces import IIdentifier, IAuthenticator, IChallenger
from repoze.who.utils import resolveDotted

import vep
from vep.utils import get_assertion_info

from repoze.who.plugins.vepauth.tokenmanager import SignedTokenManager
from repoze.who.plugins.vepauth.noncemanager import NonceManager
from repoze.who.plugins.vepauth.utils import (parse_authz_header,
                                              check_mac_signature)


class VEPAuthPlugin(object):
    """Plugin to implement VEP-Auth in Repoze.who.

    This class provides an IIdentifier, IChallenger and IAuthenticator
    implementation of repoze.who.  Authentication is based on exchanging
    a VEP assertion for a short-lived session token, which is then used
    to sign requests using HTTP MAC Authentication.

    The class takes different parameters when instanciated:

    :param audiences: the intended audiences for the browserid assertions
    :param token_url: the url to use to retrieve the token. (defaults to
                      /request_token)
    :param token_manager: the class to make and parse tokens
    :param verifier: the vep.verifiers verifier to use
    :param nonce_timeout: the timeout for the nonce (defaults to 5mn) which
                          will be used in the MAC authentication flow

    """

    implements(IIdentifier, IChallenger, IAuthenticator)

    def __init__(self, audiences, token_url=None, token_manager=None,
                 verifier=None, nonce_timeout=None):
        if isinstance(audiences, basestring):
            raise ValueError("\"audiences\" must be a list of strings")
        # Fill in default values for any unspecified arguments.
        # I'm not declaring defaults on the arguments themselves because
        # we would then have to duplicate those defaults into make_plugin.
        if token_url is None:
            token_url = "/request_token"
        if token_manager is None:
            token_manager = SignedTokenManager()
        if verifier is None:
            verifier = vep.RemoteVerifier()
        if nonce_timeout is None:
            nonce_timeout = 60
        # Now we can initialize.
        self.audiences = audiences
        if audiences:
            audience_patterns = map(self._compile_audience_pattern, audiences)
            self._audience_patterns = audience_patterns
        self.token_url = token_url
        self.token_manager = token_manager
        self.verifier = verifier
        try:
            token_timeout = token_manager.timeout
        except AttributeError:
            token_timeout = None
        self.nonce_timeout = nonce_timeout
        self.nonce_manager = NonceManager(nonce_timeout, token_timeout)

    def identify(self, environ):
        """Extract the authentication info from the request.

        If this is a request to the token-provisioning URL then we extract
        a BrowserID assertion and exchange it for a new session token, setting
        environ["repoze.who.application"] so we can pass the token details
        back to the caller.

        For all other URLs, we extract the MAC params from the Authorization
        header and return those as the identity.
        """
        request = Request(environ)
        if self._is_request_to_token_url(request):
            return self._process_vep_assertion(request)
        return self._identify_mac(request)

    def remember(self, environ, identity):
        """Remember the user's identity.

        This is a no-op for this plugin; the client is supposed to remember
        the provisioned MAC credentials and re-use them for subsequent
        requests.
        """
        return []

    def forget(self, environ, identity):
        """Forget the user's identity.

        This simply issues a new WWW-Authenticate challenge, which should
        cause the client to forget any previously-provisioned credentials.
        """
        challenge = "MAC+BrowserID url=\"%s\"" % (self.token_url,)
        return [("WWW-Authenticate", challenge)]

    def challenge(self, environ, status, app_headers=(), forget_headers=()):
        """Challenge the user for credentials.

        This simply sends a 401 response using the WWW-Authenticate field
        as constructed by forget().
        """
        resp = Response()
        resp.status = 401
        resp.headers = self.forget(environ, {})
        for headers in (app_headers, forget_headers):
            for name, value in headers:
                resp.headers[name] = value
        resp.content_type = "text/plain"
        resp.body = "Unauthorized"
        return resp

    def authenticate(self, environ, identity):
        """Authenticate the extracted identity.

        The identity must be a set of MAC auth credentials extracted from
        the request.  This method checks the MAC signature, and if valid
        extracts the user metadata from the token.
        """
        request = Request(environ)
        assert not self._is_request_to_token_url(request)
        return self._authenticate_mac(request, identity)

    #
    #  Methods for exchanging an assertion for a MAC session token.
    #

    def _process_vep_assertion(self, request):
        """Exhange a VEP assertion for some session credentials.

        This  method extracts a submitted VEP assertion, validates it and
        establishes a new session token and secret.  These are returned
        to the user so that they can sign subsequent requests as belonging
        to this session.
        """
        # Make sure they're using a GET request.
        if request.method != "GET":
            resp = Response()
            resp.status = 405
            resp.content_type = "text/plain"
            resp.body = "token requests must get GET"
            request.environ["repoze.who.application"] = resp
            return None
        # Make sure they're sending an Authorization header.
        if not request.authorization:
            msg = "you must provide an authorization header"
            return self._respond_unauthorized(request, msg)
        # Grab the assertion from the Authorization header.
        scheme, assertion = request.authorization
        if scheme.lower() != "browser-id":
            msg = "The auth scheme \"%s\" is not supported" % (scheme,)
            return self._respond_bad_request(request, msg.encode("utf8"))
        # Extract the audience, so we can check against wildcards.
        try:
            audience = get_assertion_info(assertion)["audience"]
        except (ValueError, KeyError):
            return self._respond_bad_request(request, "invalid assertion")
        if not self._check_audience(request, audience):
            msg = "The audience \"%s\" is not acceptable" % (audience,)
            return self._respond_bad_request(request, msg.encode("utf8"))
        # Verify the assertion and find out who they are.
        try:
            data = self.verifier.verify(assertion)
        except Exception, e:
            msg = "Invalid BrowserID assertion: " + str(e)
            return self._respond_bad_request(request, msg)
        # OK, we can go ahead and issue a token.
        token, secret, extra = self.token_manager.make_token(request, data)

        if token is None:
            msg = "that email address is not recognised"
            return self._respond_unauthorized(request, msg)
        resp = Response()
        resp.status = 200
        resp.content_type = "application/json"

        body = {
            "id": token,
            "key": secret,
            "algorithm": "hmac-sha-1",
        }

        if extra is not None:
            body.update(extra)

        resp.body = json.dumps(body)
        request.environ["repoze.who.application"] = resp

    def _check_audience(self, request, audience):
        """Check that the audience is valid according to our configuration.

        This function uses the configured list of valid audience patterns to
        verify the given audience.  If no audience values have been configured
        then it matches against the Host header from the request.
        """
        if not self.audiences:
            return audience == request.host_url
        for audience_pattern in self._audience_patterns:
            if audience_pattern.match(audience):
                return True
        return False

    def _compile_audience_pattern(self, pattern):
        """Compile a glob-style audience pattern into a regular expression."""
        re_pattern = fnmatch.translate(pattern)
        if "://" not in pattern:
            re_pattern = "[a-z]+://" + re_pattern
        return re.compile(re_pattern)

    #
    #  Methods for MAC Authentication once the assertion has been verified.
    #

    def _identify_mac(self, request):
        """Parse, validate and return the request's MAC parameters.

        This method grabs the MAC credentials from the Authorization header
        and performs some sanity-checks.  If the credentials are missing or
        malformed then it returns None; if they're ok then they are returned
        in a dict.

        Note that this method does *not* validate the MAC signature.
        """
        params = parse_authz_header(request, None)
        if params is None:
            return None
        if params.get("scheme") != "MAC":
            return None
        # Check that various parameters are as expected.
        token = params.get("id")
        if token is None:
            msg = "missing MAC id"
            return self._respond_unauthorized(request, msg)
        # Check the timestamp and nonce for freshness or reuse.
        # TODO: the spec requires us to adjust for per-client clock skew.
        try:
            timestamp = int(params["ts"])
        except (KeyError, ValueError):
            msg = "missing or malformed MAC timestamp"
            return self._respond_unauthorized(request, msg)
        nonce = params.get("nonce")
        if nonce is None:
            msg = "missing MAC nonce"
            return self._respond_unauthorized(request, msg)
        if not self.nonce_manager.is_fresh(token, timestamp, nonce):
            msg = "MAC has stale token or nonce"
            return self._respond_unauthorized(request, msg)
        # OK, they seem like sensible MAC paramters.
        return params

    def _authenticate_mac(self, request, identity):
        # Check that these are MAC auth credentials.
        # They may not be if we're using multiple auth methods.
        if identity.get("scheme") != "MAC":
            return None
        token = identity["id"]
        # Decode the token.
        try:
            data, secret = self.token_manager.parse_token(token)
        except ValueError:
            msg = "invalid MAC id"
            return self._respond_unauthorized(request, msg)
        # Check the MAC signature.
        if not check_mac_signature(request, secret, identity):
            msg = "invalid MAC signature"
            return self._respond_unauthorized(request, msg)
        # Store the nonce to avoid re-use.
        # We do this *after* successul auth to avoid DOS attacks.
        nonce = identity["nonce"]
        timestamp = int(identity["ts"])
        self.nonce_manager.add_nonce(token, timestamp, nonce)
        # Update the identity with the data from the token.
        identity.update(data)
        return identity["repoze.who.userid"]

    #
    #  Misc helper methods.
    #

    def _respond_bad_request(self, request, message="Bad Request"):
        """Generate a "400 Bad Request" error response."""
        resp = Response()
        resp.status = 400
        resp.body = message
        request.environ["repoze.who.application"] = resp
        return None

    def _respond_unauthorized(self, request, message="Unauthorized"):
        """Generate a "401 Unauthorized" error response."""
        resp = Response()
        resp.status = 401
        resp.headers = self.forget(request.environ, {})
        resp.content_type = "text/plain"
        resp.body = message
        request.environ["repoze.who.application"] = resp
        return None

    def _is_request_to_token_url(self, request):
        """Check if given request is to the token-provisioning URL."""
        if not self.token_url:
            return False

        if self.token_url == request.path:
            return True

        request.match(self.token_url)

        if request.matchdict:
            return True

        return False


def make_plugin(audiences=None, token_url=None, nonce_timeout=None, **kwds):
    """Make a VEPAuthPlugin using values from a .ini config file.

    This is a helper function for loading a VEPAuthPlugin via the
    repoze.who .ini config file system. It converts its arguments from
    strings to the appropriate type then passes them on to the plugin.
    """
    # You *must* specify the "audiences" parameter since it's integral
    # to the security of the protocol.  If you want it set to None to
    # allow checking based on HTTP_HOST, set it to the empty string.
    if audiences is None:
        raise ValueError('You must specify the "audiences" parameter')
    if not audiences:
        audiences = None
    elif isinstance(audiences, basestring):
        audiences = audiences.split()
    # Load the token manager, possibly from a class+args.
    token_manager = _load_from_callable("token_manager", kwds)
    # Load the VEP verifier, possibly from a class+args.
    # Assume "urlopen" is a dotted-name of a callable.
    verifier = _load_from_callable("verifier", kwds, converters={
        "urlopen": resolveDotted
    })
    # If there are any kwd args left over, that's an error.
    for unknown_kwd in kwds:
        raise TypeError("unknown keyword argument: %s" % unknown_kwd)
    plugin = VEPAuthPlugin(audiences, token_url, token_manager, verifier,
                           nonce_timeout)
    return plugin


def _load_from_callable(name, kwds, converters={}):
    """Load a plugin argument from dotted python name of callable.

    This function is a helper to load and possibly instanciate an argument
    to the plugin.  It grabs the value from the dotted python name found in
    kwds[name].  If this is a callable, it it looks for arguments of the form
    kwds[name_*] and calls the object with them.
    """
    # See if we actually have the named object.
    dotted_name = kwds.pop(name, None)
    if dotted_name is None:
        return None
    obj = resolveDotted(dotted_name)
    # Extract any arguments for the callable.
    obj_kwds = {}
    prefix = name + "_"
    for key in kwds.keys():
        if key.startswith(prefix):
            obj_kwds[key[len(prefix):]] = kwds.pop(key)
    # To any type conversion on the arguments.
    for key, value in obj_kwds.iteritems():
        converter = converters.get(key)
        if converter is not None:
            obj_kwds[key] = converter(value)
    # Call it if callable.
    if callable(obj):
        obj = obj(**obj_kwds)
    elif obj_kwds:
        raise ValueError("arguments provided for non-callable %r" % (name,))
    return obj


class Request(Request_):
    def __init__(self, *args, **kwargs):
        super(Request, self).__init__(*args, **kwargs)
        self.matchdict = {}

    def match(self, route):
        """Set the matchdict parameter given a route"""
        regexp = re.sub(r"\{([a-zA-Z][^\}]*)\}", r"(?P<\1>([^}]+))", route)
        r = re.match(regexp, self.path)
        if r is not None:
            self.matchdict = r.groupdict()
